File: tests/python/test_diag.py
"""
诊断模块测试 — 通过 nanobind 绑定测试 C++ DiagManager
"""
import threading

from sultan_core import diag
from tests.python.test_runner import TestResult, assert_eq, assert_true, run_test


def test_emit_to_buffer() -> None:
    """emit 三级别消息 -> snapshot 返回正确消息"""
    mgr = diag.get_manager()
    mgr.snapshot()  # 清空

    diag._test_emit(diag.Level.INFO, "parse", "info msg")
    diag._test_emit(diag.Level.WARN, "parse", "warn msg")
    diag._test_emit(diag.Level.ERROR, "merge", "error msg")

    msgs = mgr.snapshot()
    assert_eq(len(msgs), 3, "消息总数")
    assert_eq(msgs[0].level, diag.Level.INFO)
    assert_eq(msgs[0].category, "parse")
    assert_eq(msgs[0].message, "info msg")
    assert_eq(msgs[1].level, diag.Level.WARN)
    assert_eq(msgs[2].level, diag.Level.ERROR)
    assert_eq(msgs[2].category, "merge")


def test_snapshot_by_category() -> None:
    """按类别拉取，另一类别保留"""
    mgr = diag.get_manager()
    mgr.snapshot()

    diag._test_emit(diag.Level.WARN, "parse", "p1")
    diag._test_emit(diag.Level.WARN, "parse", "p2")
    diag._test_emit(diag.Level.ERROR, "merge", "m1")

    parse_msgs = mgr.snapshot(["parse"])
    assert_eq(len(parse_msgs), 2, "parse 类别消息数")

    remaining = mgr.snapshot()
    assert_eq(len(remaining), 1, "剩余消息数")
    assert_eq(remaining[0].category, "merge")


def test_snapshot_clears() -> None:
    """snapshot 后再次调用返回空"""
    mgr = diag.get_manager()
    mgr.snapshot()

    diag._test_emit(diag.Level.WARN, "test", "msg")
    first = mgr.snapshot()
    assert_eq(len(first), 1)

    second = mgr.snapshot()
    assert_eq(len(second), 0, "第二次 snapshot 应为空")


def test_notify_with_callback() -> None:
    """注册回调 + notify=true -> 回调触发，消息不在 buffer"""
    mgr = diag.get_manager()
    mgr.snapshot()

    received: list[tuple[diag.Level, str, str]] = []

    def on_diag(level: diag.Level, category: str, msg: str) -> None:
        received.append((level, category, msg))

    mgr.set_callback(on_diag)
    try:
        diag._test_emit(diag.Level.ERROR, "merge", "critical", notify=True)

        assert_eq(len(received), 1, "回调应触发一次")
        assert_eq(received[0][0], diag.Level.ERROR)
        assert_eq(received[0][1], "merge")
        assert_eq(received[0][2], "critical")

        msgs = mgr.snapshot()
        assert_eq(len(msgs), 0, "notify=true 的消息不应进 buffer")
    finally:
        mgr.clear_callback()


def test_notify_without_callback() -> None:
    """无回调 + notify=true -> 消息回退进 buffer"""
    mgr = diag.get_manager()
    mgr.snapshot()
    mgr.clear_callback()

    diag._test_emit(diag.Level.ERROR, "merge", "fallback", notify=True)

    msgs = mgr.snapshot()
    assert_eq(len(msgs), 1, "应回退进 buffer")
    assert_eq(msgs[0].message, "fallback")


def test_notify_false_with_callback() -> None:
    """有回调 + notify=false -> 消息进 buffer，回调未触发"""
    mgr = diag.get_manager()
    mgr.snapshot()

    called = False

    def on_diag(level: diag.Level, category: str, msg: str) -> None:
        nonlocal called
        called = True

    mgr.set_callback(on_diag)
    try:
        diag._test_emit(diag.Level.ERROR, "merge", "normal", notify=False)

        assert_true(not called, "notify=false 不应触发回调")
        msgs = mgr.snapshot()
        assert_eq(len(msgs), 1, "应进 buffer")
    finally:
        mgr.clear_callback()


def test_clear_callback() -> None:
    """注册后清除 -> emit(notify=true) -> 消息进 buffer"""
    mgr = diag.get_manager()
    mgr.snapshot()

    called = False

    def on_diag(level: diag.Level, category: str, msg: str) -> None:
        nonlocal called
        called = True

    mgr.set_callback(on_diag)
    mgr.clear_callback()

    diag._test_emit(diag.Level.ERROR, "merge", "after clear", notify=True)

    assert_true(not called, "清除回调后不应触发")
    msgs = mgr.snapshot()
    assert_eq(len(msgs), 1, "应回退进 buffer")


def test_thread_safety() -> None:
    """8 线程各 push 100 条 -> snapshot 总数 800"""
    mgr = diag.get_manager()
    mgr.snapshot()

    n_threads = 8
    msgs_per_thread = 100
    barrier = threading.Barrier(n_threads)

    def worker(tid: int) -> None:
        barrier.wait()
        for i in range(msgs_per_thread):
            diag._test_emit(diag.Level.WARN, "thread", f"t{tid}-{i}")

    threads = [threading.Thread(target=worker, args=(t,))
               for t in range(n_threads)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()

    msgs = mgr.snapshot()
    assert_eq(len(msgs), n_threads * msgs_per_thread, "并发消息总数")


def test_message_repr() -> None:
    """Message.__repr__ 格式正确"""
    mgr = diag.get_manager()
    mgr.snapshot()

    diag._test_emit(diag.Level.WARN, "parse", "test repr")
    msgs = mgr.snapshot()
    r = repr(msgs[0])
    assert_true("WARN" in r, "__repr__ 应包含级别")
    assert_true("parse" in r, "__repr__ 应包含类别")


def run_all(result: TestResult) -> None:
    run_test("diag_emit_to_buffer", test_emit_to_buffer, result)
    run_test("diag_snapshot_by_category", test_snapshot_by_category, result)
    run_test("diag_snapshot_clears", test_snapshot_clears, result)
    run_test("diag_notify_with_callback", test_notify_with_callback, result)
    run_test("diag_notify_without_callback", test_notify_without_callback, result)
    run_test("diag_notify_false_with_callback", test_notify_false_with_callback, result)
    run_test("diag_clear_callback", test_clear_callback, result)
    run_test("diag_thread_safety", test_thread_safety, result)
    run_test("diag_message_repr", test_message_repr, result)

File: tests/cpp/test_diag.cpp
#include <catch2/catch_test_macros.hpp>

#include <algorithm>
#include <atomic>
#include <string>
#include <thread>
#include <vector>

#include "diag.h"

using namespace sultan;

static DiagManager make_manager() { return DiagManager{}; }

TEST_CASE("diag: emit all levels to buffer") {
    auto mgr = make_manager();

    mgr.info("parse", "info msg");
    mgr.warn("parse", "warn msg");
    mgr.error("merge", "error msg");

    auto msgs = mgr.snapshot();
    REQUIRE(msgs.size() == 3);

    std::sort(msgs.begin(), msgs.end(),
              [](auto& a, auto& b) { return a.level < b.level; });

    REQUIRE(msgs[0].level == DiagLevel::Info);
    REQUIRE(msgs[0].category == "parse");
    REQUIRE(msgs[0].message == "info msg");

    REQUIRE(msgs[1].level == DiagLevel::Warn);
    REQUIRE(msgs[1].category == "parse");

    REQUIRE(msgs[2].level == DiagLevel::Error);
    REQUIRE(msgs[2].category == "merge");
}

TEST_CASE("diag: snapshot filters by category") {
    auto mgr = make_manager();

    mgr.warn("parse", "p1");
    mgr.warn("parse", "p2");
    mgr.error("merge", "m1");

    auto parse_msgs = mgr.snapshot({"parse"});
    REQUIRE(parse_msgs.size() == 2);
    REQUIRE(parse_msgs[0].message == "p1");
    REQUIRE(parse_msgs[1].message == "p2");

    auto remaining = mgr.snapshot();
    REQUIRE(remaining.size() == 1);
    REQUIRE(remaining[0].category == "merge");
}

TEST_CASE("diag: snapshot drains buffer") {
    auto mgr = make_manager();

    mgr.warn("test", "msg");
    auto first = mgr.snapshot();
    REQUIRE(first.size() == 1);

    auto second = mgr.snapshot();
    REQUIRE(second.empty());
}

TEST_CASE("diag: notify with callback bypasses buffer") {
    auto mgr = make_manager();

    DiagLevel cb_level{};
    std::string cb_cat, cb_msg;
    mgr.set_callback([&](DiagLevel l, const std::string& c,
                         const std::string& m) {
        cb_level = l;
        cb_cat = c;
        cb_msg = m;
    });

    mgr.error("merge", "critical", true);

    REQUIRE(cb_level == DiagLevel::Error);
    REQUIRE(cb_cat == "merge");
    REQUIRE(cb_msg == "critical");

    auto msgs = mgr.snapshot();
    REQUIRE(msgs.empty());
}

TEST_CASE("diag: notify without callback falls back to buffer") {
    auto mgr = make_manager();

    mgr.error("merge", "critical", true);

    auto msgs = mgr.snapshot();
    REQUIRE(msgs.size() == 1);
    REQUIRE(msgs[0].message == "critical");
}

TEST_CASE("diag: notify=false with callback goes to buffer") {
    auto mgr = make_manager();

    bool called = false;
    mgr.set_callback([&](DiagLevel, const std::string&, const std::string&) {
        called = true;
    });

    mgr.error("merge", "normal", false);

    REQUIRE_FALSE(called);
    auto msgs = mgr.snapshot();
    REQUIRE(msgs.size() == 1);
}

TEST_CASE("diag: clear callback then notify falls back") {
    auto mgr = make_manager();

    bool called = false;
    mgr.set_callback([&](DiagLevel, const std::string&, const std::string&) {
        called = true;
    });
    mgr.clear_callback();

    mgr.error("merge", "after clear", true);

    REQUIRE_FALSE(called);
    auto msgs = mgr.snapshot();
    REQUIRE(msgs.size() == 1);
}

TEST_CASE("diag: thread safety") {
    auto mgr = make_manager();

    const int n_threads = 8;
    const int msgs_per_thread = 100;

    std::vector<std::thread> threads;
    std::atomic<int> ready{0};

    for (int t = 0; t < n_threads; ++t) {
        threads.emplace_back([&mgr, &ready, t, n_threads, msgs_per_thread]() {
            ready.fetch_add(1);
            while (ready.load() < n_threads) {}
            for (int i = 0; i < msgs_per_thread; ++i) {
                mgr.warn("thread",
                          "t" + std::to_string(t) + "-" + std::to_string(i));
            }
        });
    }

    for (auto& th : threads) th.join();

    auto msgs = mgr.snapshot();
    REQUIRE(msgs.size() == n_threads * msgs_per_thread);
}
